contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,493 @@
+                                    "type": {
+                                        "enum": [
+                                            "one",
+                                            "two"
+                                        ]
+                                    }
+                                },
+                                "required": [
+                                    "type"
+                                ],
+                                "additionalProperties": false
+                            }
+                        ]
+                    },
+                    "additionalProperties": false,
+                    "definitions": {
+                        "zeta": {
+                            "description": "Every zeta needs to have foo, bar and baz",
+                            "type": "object",
+                            "properties": {
+                                "foo": {
+                                    "type": "string"
+                                },
+                                "bar": {
+                                    "type": "number"
+                                },
+                                "baz": {
+                                    "type": "boolean"
+                                }
+                            },
+                            "required": [
+                                "foo",
+                                "bar",
+                                "baz"
+                            ],
+                            "additionalProperties": false
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+## <a name="behavior"></a>Validation Behavior
+
+The schema will be described in the `CustomResourceDefinitionSpec`. The validation will be carried out using the [go-openapi validation library](https://github.com/go-openapi/validate).
+ 
+While creating/updating the CR, the metadata is first validated. To validate the CR against the spec in the CRD, we _must_ have server-side validation and we _can_ have client-side validation.
+
+### <a name="metadata"></a>Metadata
+
+ObjectMeta and TypeMeta are implicitly specified. They do not have to be added to the JSON-Schema of a CRD. The validation already happens today as part of the apiextensions-apiserver REST handlers.
+
+### <a name="server-side"></a>Server-Side Validation
+
+The server-side validation is carried out after sending the request to the apiextensions-apiserver, i.e. inside the CREATE and UPDATE handlers for CRs.
+
+We do a schema pass there using the https://github.com/go-openapi/validate validator with the provided schema in the corresponding CRD. Validation errors are returned to the caller as for native resources.
+
+JSON-Schema also allows us to reject additional fields that are not defined in the schema and only allow the fields that are specified. This can be achieved by using `"additionalProperties": false` in the schema.
```

Comment:
To elaborate, my interpretation of the chain of comments @bgrant0607 linked is: Yes, there is danger in allowing CRD authors to set `additionalProperties: false` because it breaks version skew (new client sends new optional field to old server). At most, the client should be able to opt-in to "reject unrecognized fields" on a per-request basis. However, there is currently no such standard query param (AFAIK), so we should not allow CRD authors to set `additionalProperties: false`.